app/src/main/java/com/tari/android/wallet/infrastructure/backup/WalletRestoration.kt

```diff
@@ -0,0 +1,44 @@
+package com.tari.android.wallet.infrastructure.backup
+
+import com.orhanobut.logger.Logger
+import com.tari.android.wallet.infrastructure.backup.compress.CompressionMethod
+import com.tari.android.wallet.infrastructure.backup.storage.BackupStorage
+import java.io.File
+
+class WalletRestoration(
+    private val storage: BackupStorage,
+    private val cd: File,
+    private val compressionMethod: CompressionMethod
+) {
+
+    fun run() {
+        require(storage.doesBackupExist()) { "Backup does not exist" }
+        val archive = File(cd, "restore_archive_${System.currentTimeMillis()}.zip")
+        val restoreDir = File(cd, "restore/").apply { mkdir() }
+        try {
+            storage.downloadBackup(archive)
+            compressionMethod.unpack(archive, restoreDir)
+            archive.delete()
+            BACKUP_FILES.forEach { File(restoreDir, it).apply { copyTo(File(cd, name), overwrite = true) } }
+            restoreDir.deleteRecursively()
+        } catch (e: Exception) {
+            Logger.e(e, "Exception occurred during restoration")
+            archive.delete()
+            restoreDir.deleteRecursively()
+            throw e
+        }
+    }
+
+}
+
+class WalletRestorationFactory(
```

Comment:
Do we really need a factory here?

Reply:
yeah, otherwise it wouldn't exist

the problem is that `BackupStorage` instance is acquired only after a successful google sign-in, so it exists to provide a restoration object when given a `BackupStorage` instance. this way encapsulation of unnecessary for a fragment things like `compressionMethod`, `filesDirectory` and specific `WalletRestoration` instance are achieved while having an ability to get this particular abstraction when it is possible